macros/src/main/scala/gestalt/macros/Monadless.scala

```diff
@@ -83,120 +92,159 @@ object Transformer {
+              Some(If(cond, ifTrue, ifFalse))
+            case None =>
+              Some(If(cond, ifTrue, ifFalse))
+          }
+
+        case _ => None
+      }
+
+      def unapply(ifTrue: tpd.Tree, ifFalse: tpd.Tree): Option[(tpd.Tree, tpd.Tree)] =
+        (ifTrue, ifFalse) match {
+          case (Transform(ifTrue), Transform(ifFalse)) =>
+            Some((ifTrue, ifFalse))
+          case (Transform(ifTrue), ifFalse) =>
+            val elsep = Apply(Resolve.apply(tree.pos), List(ifFalse))
+            Some((ifTrue, elsep))
+          case (ifTrue, Transform(ifFalse)) =>
+            val ifp = Apply(Resolve.apply(tree.pos), List(ifTrue))
+            Some((ifp, ifFalse))
+          case (ifTrue, ifFalse) =>
+            None
         }
     }
 
     object Transform {
 
-      def apply(tree: Tree): Tree =
+      def apply(tree: tpd.Tree): tpd.Tree =
         unapply(tree).getOrElse(tree)
 
-      def unapply(tree: Tree): Option[TermTree] = tree match {
+      def unapply(tree: tpd.Tree): Option[tpd.Tree] = tree match {
         case PureTree(tree) => None
 
         case Ascribe(tree, _) => unapply(tree)
 
-        case Block(trees) if trees.size > 1 => TransformBlock.unapply(trees.toList)
-
-        case If(Transform(monad), ifTrue, ifFalse) =>
-          val name = fresh()
-          val body = If(Ident(name), ifTrue, ifFalse)
-          Some(Nest(monad, name, body))
-
-        case If(cond, ifTrue, Some(ifFalse)) =>
-          (ifTrue, ifFalse) match {
-            case (Transform(ifTrue), Transform(ifFalse)) =>
-              Some(If(cond, ifTrue, Some(ifFalse)))
-            case (Transform(ifTrue), ifFalse) =>
-              val elsep = Apply(Resolve.apply(tree.pos), List(ifFalse))
-              Some(If(cond, ifTrue, Some(elsep)))
-            case (ifTrue, Some(Transform(ifFalse))) =>
-              val ifp = Apply(Resolve.apply(tree.pos), List(ifTrue))
-              Some(If(cond, ifp, Some(ifFalse)))
-            case (ifTrue, ifFalse) =>
-              None
-          }
+        case block @ Block(stats, expr) => TransformBlock.unapply(stats.toList :+ expr)(block.tpe)
 
+        case If(cond, ifTrue, ifFalse) => TransformIf.unapply(tree)
 
         case q"$fun[$tp]($v)" if isUnlift(fun.tpe) => Some(v)
 
-        case tree: Tree =>
-          val unlifts = collection.mutable.ListBuffer.empty[(TermTree, String, TypeTree)]
-          val newTree: TermTree =
+        case tree =>
+          val unlifts = collection.mutable.ListBuffer.empty[(tpd.Tree, Symbol, tpd.Tree)]
+          val newTree =
             transform(tree) {
-              case q"$fun[$tp]($v)" if isUnlift(fun.tpe) =>
+              case tree @ q"$fun[$tp]($v)" if isUnlift(fun.tpe) =>
                 val name = fresh()
-                val splice = wrap(tp)
-                unlifts += ((v, name, splice))
-                Ident(name)
-            }.asInstanceOf[TermTree]
+                val dummy = newValSymbol(name, tree.tpe)
```

Comment:
For this macro, seems we have no need to do so explicitly. The assumption is that user create symbols against `ctx.owner`, and in the construction `Function`, we automatically change subtree's owner from `ctx.owner` to `meth`.

But for definition trees destructed from the macro arguments, this scheme doesn't work. Two choices:

- in constructors, allow subtree to provide their owner explicitly, which will be used in `changeOwner` over `ctx.owner`.
- Just inspect the typed tree to find out the `owner` by code, then call `changeOwner(owner, meth)`.